oauth2_http/java/com/google/auth/oauth2/AgentIdentityUtils.java

```diff
@@ -0,0 +1,335 @@
+    }
+
+    POLLING_INTERVALS = Collections.unmodifiableList(intervals);
+  }
+
+  // Interface to allow mocking System.getenv for tests without exposing it publicly.
+  interface EnvReader {
+    String getEnv(String name);
+  }
+
+  private static EnvReader envReader = System::getenv;
+
+  /**
+   * Internal interface to allow mocking time and sleep for tests. This is used to prevent tests
+   * from running for long periods of time when polling is involved.
+   */
+  @VisibleForTesting
+  interface TimeService {
+    long currentTimeMillis();
+
+    void sleep(long millis) throws InterruptedException;
+  }
+
+  private static TimeService timeService =
+      new TimeService() {
+        @Override
+        public long currentTimeMillis() {
+          return System.currentTimeMillis();
+        }
+
+        @Override
+        public void sleep(long millis) throws InterruptedException {
+          Thread.sleep(millis);
+        }
+      };
+
+  private AgentIdentityUtils() {}
+
+  /**
+   * Gets the Agent Identity certificate if certificate is available and agent token sharing is not
+   * disabled.
+   *
+   * @return The X509Certificate if found and Agent Identities are enabled, null otherwise.
+   * @throws IOException If there is an error reading the certificate file after retries.
+   */
+  static X509Certificate getAgentIdentityCertificate() throws IOException {
+    if (isOptedOut()) {
+      return null;
+    }
+
+    String certConfigPath = envReader.getEnv(GOOGLE_API_CERTIFICATE_CONFIG);
+    if (Strings.isNullOrEmpty(certConfigPath)) {
+      return null;
+    }
+
+    String certPath = getCertificatePathWithRetry(certConfigPath);
+    return parseCertificate(certPath);
+  }
+
+  /**
+   * Checks if Agent Identity token sharing is disabled via an environment variable.
+   *
+   * @return {@code true} if the {@link #GOOGLE_API_PREVENT_AGENT_TOKEN_SHARING_FOR_GCP_SERVICES}
+   *     variable is set to {@code "false"}, otherwise returns {@code false}.
+   */
+  private static boolean isOptedOut() {
+    String optOut = envReader.getEnv(GOOGLE_API_PREVENT_AGENT_TOKEN_SHARING_FOR_GCP_SERVICES);
+    return optOut != null && "false".equalsIgnoreCase(optOut);
+  }
+
+  /**
+   * Polls for the certificate config file and the certificate file it references, and returns the
+   * certificate's path.
+   *
+   * <p>This method will retry for a total of {@link #TOTAL_TIMEOUT_MS} milliseconds before failing.
+   *
+   * @param certConfigPath The path to the certificate configuration JSON file.
+   * @return The path to the certificate file extracted from the config.
+   * @throws IOException If the files cannot be found after the timeout, or if the thread is
+   *     interrupted while waiting.
+   */
+  private static String getCertificatePathWithRetry(String certConfigPath) throws IOException {
+    boolean warned = false;
+
+    // Deterministic polling loop based on pre-calculated intervals.
+    for (long sleepInterval : POLLING_INTERVALS) {
```

Comment:
I'm concerned about this given that this is to be opt-out and not opt-in. By default, customers may now experience up to 30s of busy waiting just because the filesystem on certain environments can't be loaded fast enough. This applies for customers who don't touch Agent Identities.

IIUC the code from above, the only way to to know if Agent Identities shouldn't be used:
1. Load the cert file (up to 30s of busy waiting)
2. Then load and parse the cert file to check the SAN

OR 
Require users to explicitly add the new env var to their workloads + envs

Can you let me know if my understanding is correct? If this is the case that the cert has to be loaded, then I feel that opt-in would be better.